src/V3Param.cpp

```diff
@@ -63,908 +64,980 @@
+                if (varp->isIO() || varp->isIfaceRef() || varp->isParam()) {
+                    origNodep = varp->clonep();
                 }
+            } else if (AstParamTypeDType* const dtypep = VN_CAST(stmtp, ParamTypeDType)) {
+                if (dtypep->isGParam()) origNodep = dtypep->clonep();
             }
+            if (origNodep) m_clonedModPinMap[origNodep] = stmtp;
         }
     }
-    const ModInfo* moduleFindOrClone(AstNodeModule* srcModp, AstNode* cellp, AstPin* paramsp,
-                                     const string& newname, const IfaceRefRefs& ifaceRefRefs) {
-        // Already made this flavor?
-        auto it = m_modNameMap.find(newname);
-        if (it != m_modNameMap.end()) {
-            UINFO(4, "     De-parameterize to old: " << it->second.m_modp << endl);
-        } else {
-            deepCloneModule(srcModp, cellp, paramsp, newname, ifaceRefRefs);
-            it = m_modNameMap.find(newname);
-            UASSERT(it != m_modNameMap.end(), "should find just-made module");
+    void relinkPins(DeparamedModInfo* modInfo, AstPin* pinsp) {
+        for (AstPin* pinp = pinsp; pinp; pinp = VN_AS(pinp->nextp(), Pin)) {
+            if (AstVar* const varp = pinp->modVarp()) {
+                const auto newVarp = modInfo->findPinMap(varp);
+                UASSERT_OBJ(newVarp, pinp, "Couldn't find pin in clone list");
+                pinp->modVarp(VN_AS(newVarp, Var));
+            } else if (AstParamTypeDType* const typep = pinp->modPTypep()) {
+                const auto newTypep = modInfo->findPinMap(typep);
+                UASSERT_OBJ(newTypep, pinp, "Couldn't find pin in clone list");
+                pinp->modPTypep(VN_AS(newTypep, ParamTypeDType));
+            } else {
+                pinp->v3fatalSrc("Not linked?");
+            }
         }
-        const ModInfo* const modInfop = &(it->second);
-        return modInfop;
     }
-
-    void convertToStringp(AstNode* nodep) {
-        // Should be called on values of parameters of type string to convert them
-        // to properly typed string constants.
-        // Has no effect if the value is not a string constant.
-        AstConst* const constp = VN_CAST(nodep, Const);
-        // Check if it wasn't already converted
-        if (constp && !constp->num().isString()) {
-            constp->replaceWith(
-                new AstConst{constp->fileline(), AstConst::String{}, constp->num().toString()});
-            constp->deleteTree();
+    bool updateClonedModInfo(ModInfo* modInfop, AstNodeModule* clonedModp, ModParamSet* paramsp) {
+        AstNodeModule* const modp = modInfop->originalModp();
+        string suffix;
+        if (!modInfop->hierBlock()) {
+            suffix = "__parameterized" + std::to_string(modInfop->nextParamModIndex());
+        } else {
+            // uint32_t hash = static_cast<uint32_t>(ModParamSet::Hash()(paramsp));
+            suffix = "_hierblk" + std::to_string(modInfop->nextParamModIndex());
```

Comment:
Maybe it's better to assign a hash-based name for hierarchical blocks because it's more resistant to changed cell order.